### ISSUE TITLE
Bring examples back into the monorepo

### ISSUE THREAD

**Issue by virkt25** [strongloop/loopback-next#836]
There is a growing list of examples that need to be maintained and kept up to date with the latest version of LoopBack especially as there are breaking changes that might be made during the development of MVP. 

The example repos are kept in their own repository right now so it's easy for users to be able to clone the examples and play around with working code and this won't be possible if we move the examples to the monorepo but it would make maintenance a lot easier.

The purpose of this task is to explore options for doing this in a user friendly manner. Some avenues of exploration are:
- A CLI command that can get the relevant example from the monorepo, update package.json and stand up the example without the monorepo dependencies. Ex commands: `lb4 example getting-started` or `lb4 example log-extension`, etc. 

- A new monorepo just for examples. Not as clean but users will be able to clone the entire repo and run the individual examples by going to the appropriate section. NOTE: When I say monorepo here, I mean more a big folder of example repos and not a lerna monorepo since different examples shouldn't have dependencies on each other.

**Tasks**

 - [x] Implement `lb4 example` CLI command - see https://github.com/strongloop/loopback-next/pull/848
 - [x] Move https://github.com/strongloop/loopback4-example-getting-started to monorepo - see #869 and https://github.com/strongloop/loopback4-example-getting-started/pull/19
 - [x] Move https://github.com/strongloop/loopback4-example-log-extension to monorepo - see https://github.com/strongloop/loopback-next/pull/908
 - [x] Move https://github.com/strongloop/loopback4-example-rpc-server to monorepo - see #915
 - [x] Remove example-codehub - https://github.com/strongloop/loopback-next/pull/900/
 - [x] https://github.com/strongloop/loopback4-example-hello-world: Move it into monorepo as even simpler version than loopback4-example-getting-started. See https://github.com/strongloop/loopback-next/pull/924
 - [x] https://github.com/strongloop/loopback-next-example: Rename it to reflect it's advanced example of how multiple micro services work together. It will be great that we can evolve it with gRPC and OpenAPI based assembly. ~~See https://github.com/strongloop/loopback-next/pull/933~~
 - [x] https://github.com/strongloop/loopback4-extension-starter: DEPRECATE. ~~Move sample controller/provider/etc. source files from  to `lb4 extension` templates~~ See https://github.com/strongloop-archive/loopback4-extension-starter/commit/8e869c06ee964000d18af52518c84ad1c198939e
 - [x] DEPRECATE https://github.com/strongloop/loopback4-extension-typeorm - see https://github.com/strongloop-archive/loopback4-extension-typeorm/pull/6
 - [x] DEPRECATE https://github.com/strongloop/loopback-next-hello-world - see https://github.com/strongloop/loopback-next-hello-world/pull/26
 - [x] Update the docs: http://loopback.io/doc/en/lb4/Examples-and-tutorials.html - see https://github.com/strongloop/loopback.io/pull/595
 - [x] Write a blog post - see https://github.com/strongloop/strongloop.com/pull/110

Things to change as part of migration:
 - run `slt license` and `slt copyright`
 - mark the package as private
 - review package.json scripts - make them consistent with other monorepo packages
 - add an entry to CODEOWNERS
 - add an entry to the list of examples used by `lb4 example` 
 - update README with instructions mentioning `lb4 example`
 - move the original repo to strongloop-archive
 - remove all files from the original repo, keep README pointing to our monorepo only
 - archive the original repo (make it read-only)

See also https://github.com/strongloop/loopback-next/issues/855

**Comment by bajtos** [comment]
I personally prefer to bring all examples to the main monorepo, so that we run their tests as part of our CI for any changes we are making in loopback core.

Here are my thoughts how the CLI command could be implemented, preferably with no dependency on `git`:

 1. Start with a prompt asking the user what example to obtain. (There should be a CLI argument to skip this prompt and run in a non-iteractive mode.) The list of examples can be hard-coded for now, to keep things simple. Later, we may implement some automated solution that would fetch the list of examples from GitHub.

 2. Download a ZIP archive of loopback-next's master branch: https://github.com/strongloop/loopback-next/archive/master.zip

 3. Use [extract-zip](https://www.npmjs.com/package/extract-zip) to open the archive and extract only the files from `packages/<selected-example>`.

 4. AFAICT, there are only few things we need to fix when extracting a package from a monorepo: 
    - add some of the shared dev-dependencies from monorepo's top-level `package.json`
    - add `test/mocha.opts` from monorepo to the example package
    - fix the npm script for running mocha  - remove the reference to `mocha.opts`, since `test/mocha.opts` is loaded by default

This looks pretty doable to me :)

I am proposing the following approach:

 - Start with implementing `lb4 example` using the existing `example-codehub` package to seed the initial list of examples
 - Once `lb4 example` is landed, start moving individual example repositories to our monorepo

**Comment by bajtos** [comment]
The kick-off: #848

**Comment by raymondfeng** [comment]
@bajtos Here is my view:

 - https://github.com/strongloop/loopback4-example-hello-world (Move it into monorepo as even simpler version than loopback4-example-getting-started)
- https://github.com/strongloop/loopback-next-example (Rename it to reflect it's advanced example of how multiple micro services work together. It will be great that we can evolve it with gRPC and OpenAPI based assembly)
- https://github.com/strongloop/loopback-next-hello-world (Deprecate)

**Comment by bajtos** [comment]
>  https://github.com/strongloop/loopback4-extension-starter

@raymondfeng @kjdelisle @virkt25  I am proposing to deprecate this repository and tell our users to run `lb4 extension` instead. Any objections? Are you aware of any specific content in the extension starter that is not part of our current extension template and that should be copied over?

**Comment by virkt25** [comment]
Hi @bajtos, `loopback4-extension-starter` contains examples of different types of extension artifacts that can be used to create an extension as well as show the project structure layout. CLI tooling currently does not provide those as a template for a new project (just the folder structure). I personally think the files should be converted to templates for different artifacts that can be generated by the CLI if we are going to deprecate this repository.

**Comment by bajtos** [comment]
> Hi @bajtos, loopback4-extension-starter contains examples of different types of extension artifacts that can be used to create an extension as well as show the project structure layout. CLI tooling currently does not provide those as a template for a new project (just the folder structure). I personally think the files should be converted to templates for different artifacts that can be generated by the CLI if we are going to deprecate this repository.

I agree that those example files should be converted to templates. I am worried that creating new CLI commands for different extension artifacts would take more time than I'd like to invest into this right now (for MVP scope). Would you mind if I simply copied the example files from extension-starter to our extension template as a short-term solution allowing us to get rid of the extension-starter repository?

**Comment by virkt25** [comment]
Yep that sounds like a good solutions. I'd even say we can go as far as just having the contents as a comment in the template instead of actual code showing an example.

**Comment by bajtos** [comment]
>>> loopback4-extension-starter contains examples of different types of extension artifacts that can be used to create an extension as well as show the project structure layout. CLI tooling currently does not provide those as a template for a new project (just the folder structure). I personally think the files should be converted to templates for different artifacts that can be generated by the CLI if we are going to deprecate this repository.
>>
>> I agree that those example files should be converted to templates. I am worried that creating new CLI commands for different extension artifacts would take more time than I'd like to invest into this right now (for MVP scope). Would you @virkt25 mind if I simply copied the example files from extension-starter to our extension template as a short-term solution allowing us to get rid of the extension-starter repository?
>
> Yep that sounds like a good solutions. I'd even say we can go as far as just having the contents as a comment in the template instead of actual code showing an example.

Keeping a commented-out code is an anti-pattern to avoid. Because the comments are not checked by the compiler (and linter), the commented-out code will eventually become out of sync with the rest of the codebase and stop working.

I started the work of adding the template files as they are to our CLI template, but realized this is not a good approach. 

 - Our current extension template is pretty lightweight, it does not require the extension to depend on `@loopback/context` and `@loopback/rest`. OTOH, the starter examples do depend on these two packages. 
 - There are quite few names that are parameterized by the extension name, e.g. `ExtensionStarterBindings` namespace. While our templating system allows us to modify the scaffolded files to rename `ExtensionStarterBinding` to `${ComponentName}Bindings`, I feel it will clutter our code too much.
 - We don't have any test to verify that the scaffolded extension project actually passes `npm test`. This is is something we need to add, but I'd rather not do that as part of this story - see https://github.com/strongloop/loopback-next/issues/932

In that light, I am considering the following options:

 - Move extension-starter to our monorepo to serve as an example showing different things an extension can contribute, or:
 - Deprecate extension-starter repo (despite the reasons stated above), use example-log-extension as the example showing how to write different extension artifacts.

Please note that the current CLI template does provide some guidance on how to write different extension artifacts, see e.g. [src/decorators/README.md](https://github.com/strongloop/loopback-next/blob/afac0cfd0e738dca4903be7b0f48631daa5fd1b2/packages/cli/generators/extension/templates/src/decorators/README.md).

@raymondfeng @kjdelisle @virkt25 thoughts?

**Comment by bajtos** [comment]
> https://github.com/strongloop/loopback-next-example (Rename it to reflect it's advanced example of how multiple micro services work together. It will be great that we can evolve it with gRPC and OpenAPI based assembly)

Based on this comment and the discussion in https://github.com/strongloop/loopback-next/pull/933, I renamed the repository to https://github.com/strongloop/loopback4-example-microservices. I'd like to keep any additional changes out of scope of this already-too-large issue.

**Comment by bajtos** [comment]
> https://github.com/strongloop/loopback-next-example (Rename it to reflect it's advanced example of how multiple micro services work together. It will be great that we can evolve it with gRPC and OpenAPI based assembly)

Based on this comment and the discussion in https://github.com/strongloop/loopback-next/pull/933, I renamed the repository to https://github.com/strongloop/loopback4-example-microservices. I'd like to keep any additional changes out of scope of this already-too-large issue.

**Comment by virkt25** [comment]
> Keeping a commented-out code is an anti-pattern to avoid. Because the comments are not checked by the compiler (and linter), the commented-out code will eventually become out of sync with the rest of the codebase and stop working.

Good point about commented out code becoming outdated. 

With respect to extension-starter, I always looked at the project to serve as a template for some of the files that have a more involved setup such as `mixins`, `decorators`, etc. That said we now have a decorator factory and this project never had a complete complex decorator template. 

Only template I see that might be helpful is that for a mixin but I don’t think it’s worth keeping this project around just for that when one can get the template from `example-log-extension`.

TLDR I think it’s better to deprecate `extension-starter`.

**Comment by kjdelisle** [comment]
Let's drop extension-starter, then.

**Comment by shimks** [comment]
@bajtos I personally think we need docs for `lb4 example`, same way as all the other CLI commands are documented under `Command-line interface` tab in `loopback.io` sidebar. What do you think?

**Comment by bajtos** [comment]
@shimks Good catch! I assume you mean the following section? http://loopback.io/doc/en/lb4/Command-line-interface.html

**Comment by bajtos** [comment]
@shimks I added the documentation as you suggested, please review: https://github.com/strongloop/loopback.io/pull/614

**Comment by bajtos** [comment]
Done 🕺 🎉
